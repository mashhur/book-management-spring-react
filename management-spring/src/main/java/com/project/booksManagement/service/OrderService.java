package com.project.booksManagement.service;

import java.util.List;

import com.project.booksManagement.model.BookOrder;

public interface OrderService  {

	public BookOrder addBookOrder(BookOrder BookOrder) ;
	public BookOrder updateBookOrderStatus(BookOrder BookOrder,Integer id);
//	public void deleteBookOrder(Integer id);
	public List<BookOrder> viewAllOrders();
	public BookOrder getBookOrder(Integer id);
	public List<BookOrder> viewAllOrdersForUser(Integer id);

}
