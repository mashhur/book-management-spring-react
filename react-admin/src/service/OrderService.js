const axios = require("axios");

const BASE_URL = "http://localhost:8080/api/order";

export const getAllOrders = async () => axios.get(BASE_URL);
export const getOrder = (id) =>
	axios.get(`${BASE_URL}/${id}`);
export const updateStatus = (data, id) =>
	axios.put(`${BASE_URL}/change/${id}`, data);
