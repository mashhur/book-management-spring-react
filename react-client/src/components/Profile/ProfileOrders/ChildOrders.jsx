import React from "react";
import { AiOutlineArrowLeft } from "react-icons/ai";
import RatingComponent from "../../Ratings";

const ChildOrders = ({ orders, setchildView }) => {
	return (
		<div
			data-test="component-profile-order"
			className=" container ">
			<div
				className="overflow-scroll px-3"
				style={{ height: "80vh" }}>
				<h3 className="mb-4">
					<AiOutlineArrowLeft
						className="me-3 fs-2 hover"
						onClick={() => setchildView(false)}
					/>
					{orders.books.length} more
				</h3>
				{orders.books.map((book, index) => (
					<div className="product">
						<div className="book ">
							<img src={book.cover} />
							<div>
								<p
									className={`m-0  p-0 status ${
										orders.status == "Cancelled"
											? "text-danger"
											: ""
									}`}>
									{orders.statusMsg}
								</p>
								<h3 className="m-0 p-0">{book.title}</h3>
								<p className="author"> â‚¹{book.price}</p>
								<small className="text-muted">
									{orders.orderDate}
								</small>
							</div>
						</div>
						<div className="ratings">
							<RatingComponent book={book} />
						</div>
					</div>
				))}
			</div>
		</div>
	);
};

export default ChildOrders;
