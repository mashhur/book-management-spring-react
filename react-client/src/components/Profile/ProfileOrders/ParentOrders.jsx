import React from "react";
import RatingComponent from "../../Ratings";
import PropTypes from "prop-types";

const ParentOrders = ({ orders, setChildBooks }) => {
	const getBookName = (books) => {
		return books.length == 1
			? books[0].title
			: `${books[0].title} + ${books.length} more `;
	};

	const viewAllClickHandler = (order) => {
		console.log("setting orders child");
		console.log(order);
		setChildBooks(order);
	};
	return (
		<div
			data-test="component-profile-order-parent"
			className=" container ">
			<div
				className="overflow-scroll px-3"
				style={{ height: "80vh" }}>
				{orders.map((order, index) => (
					<div
						data-test="order-item-component"
						key={index}
						className="product">
						<div className="book ">
							<img src={order.books[0].cover} />
							<div>
								<p
									className={`m-0  p-0 status ${
										order.status == "Cancelled"
											? "text-danger"
											: ""
									}`}>
									{order.statusMsg}
								</p>
								<h3 className="m-0 p-0">
									{getBookName(order.books)}
								</h3>
								<p className="author"> â‚¹{order.price}</p>
								<small className="text-muted">
									{order.orderDate}
								</small>
							</div>
						</div>
						{order.books.length == 1 ? (
							<div className="ratings">
								<RatingComponent book={order.books[0]} />
							</div>
						) : (
							<div className="other">
								<button
									data-test="show-more-books-button"
									onClick={() => viewAllClickHandler(order)}
									className=" badge alert-warning p-2  px-3 btn btn-warning rounded-pill badge-pill">
									view {order.books.length} other books
								</button>
							</div>
						)}
					</div>
				))}
			</div>
		</div>
	);
};

ParentOrders.propType = {
	orders: PropTypes.array.isRequired,
	setChildBooks: PropTypes.func.isRequired,
};
export default ParentOrders;
