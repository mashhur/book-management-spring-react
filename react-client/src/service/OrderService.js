const axios = require("axios");

const BASE_URL = "http://localhost:8080/api/order";

class OrderService {
	async getAllOrdersFor(id) {
		return axios.get(`${BASE_URL}/u/${id}`);
	}
	async placeOrder(order) {
		return axios.post(`${BASE_URL}/add`, order);
	}
}

module.exports = new OrderService();
