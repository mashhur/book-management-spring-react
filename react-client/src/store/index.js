import { configureStore } from "@reduxjs/toolkit";
import locationReducer from "./slices/locationSlice";
import userReducer from "./slices/userSlice";

export const store = configureStore({
	reducer: {
		location: locationReducer,
		user: userReducer,
	},
});
