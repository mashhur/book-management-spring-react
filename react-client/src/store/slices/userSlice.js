import { createSlice } from "@reduxjs/toolkit";

const inituser = JSON.parse(localStorage.getItem("user"));
const user = inituser || {
	cid: "",
	fullName: "",
	phoneNumber: "",
};
const initialState = {
	auth: localStorage.getItem("auth") || false,
	user: user,
	address: {
		aid: "",
		customerName: "",
		phoneNumber: "",
		address: "",
		mstate: "",
		pincode: "",
		customerId: "",
	},
	books: [],
	cart: [],
};

export const userSlice = createSlice({
	name: "user",
	initialState,
	reducers: {
		changeAuth: (state) => {
			state.auth = true;
		},
		setUser: (state, action) => {
			state.user = action.payload;
		},
		setAddress(state, action) {
			state.address = action.payload;
		},
		addBook(state, action) {
			state.books = [];
			state.books.push(action.payload);
		},
		addBooktoCart(state, action) {
			state.cart.push(action.payload);
		},
		addtoBuy(state) {
			const cart = state.cart;
			state.books = cart;
			state.cart = [];
		},
		clearOrderDetails(state) {
			state.books = [];
			state.address = {};
		},
		removeBook(state, action) {
			state.cart.splice(action.payload, 1);
		},
		clearAuth(state) {
			state.auth = false;
			state.user = {
				cid: "",
				fullName: "",
				phoneNumber: "",
			};
		},
	},
});

export const {
	changeAuth,
	setUser,
	setAddress,
	addBook,
	clearOrderDetails,
	removeBook,
	addBooktoCart,
	addtoBuy,
	clearAuth,
} = userSlice.actions;

export default userSlice.reducer;
