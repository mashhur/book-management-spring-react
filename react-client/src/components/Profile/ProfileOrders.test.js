import React from "react";
import { mount } from "enzyme";
import { findByTestAttr } from "../../tests/testUtils";
import { Provider } from "react-redux";
import { createTestStore } from "../../store";
import Profileorders from "./ProfileOrders";
import OrderService from "../../service/OrderService";
import { setUser } from "../../store/slices/userSlice";
import moxios from "moxios";
import { DummyData } from "../../tests/dummyData";
import ParentOrders from "./ProfileOrders/ParentOrders";
import ChildOrders from "./ProfileOrders/ChildOrders";

const defaultState = {
	orders: [],
	childView: false,
	childOrders: null,
	isLoading: true,
	error: null,
};

const store = createTestStore();
const setup = () => {
	return mount(
		<Provider store={store}>
			<Profileorders />
		</Provider>
	);
};

describe("Profile orders component test casses", () => {
	beforeAll(() => {
		moxios.install();
	});

	afterAll(() => {
		moxios.uninstall();
	});
	it("renders without any error", async () => {
		const wrapper = setup();
		const component = await findByTestAttr(
			wrapper,
			"component-profile-order"
		);
		expect(component.length).toBe(1);
	});

	describe("when `getAllOrdersFor()` is called", () => {
		it("return the correct repsonse", async () => {
			await moxios.wait(() => {
				const req = moxios.requests.mostRecent();
				req.respondWith({
					status: 200,
					response: "yes",
				});
			});
			return OrderService.getAllOrdersFor().then(
				(respsonse) => {
					expect(respsonse.data).toEqual("yes");
				}
			);
		});

		//response error has been tested.
		it("return error response", async () => {
			await moxios.wait(() => {
				const req = moxios.requests.mostRecent();
				const errResp = {
					status: 500,
					response: {
						msg: "The server is unable to respond",
					},
				};
				req.reject(errResp);
			});

			return OrderService.getAllOrdersFor().catch(
				(error) => {
					// eslint-disable-next-line jest/no-conditional-expect
					expect(error.status).toEqual(500);
				}
			);
		});
	});

	describe("when an api call is made", () => {
		it("should call fn once only", async () => {
			jest
				.spyOn(React, "useEffect")
				.mockImplementationOnce((f) => f());
			jest.spyOn(OrderService, "getAllOrdersFor");
			setup();
			expect(
				OrderService.getAllOrdersFor
			).toHaveBeenCalledTimes(1);
		});

		// test loading state
		describe("while loading", () => {
			let wrapper;
			beforeEach(() => {
				const initalValue = {
					isLoading: true,
					error: null,
				};

				React.useState = jest.fn(() => [
					initalValue,
					jest.fn(),
				]);
				wrapper = setup();
			});

			it("should render loading", async () => {
				const component = await findByTestAttr(
					wrapper,
					"loading-spinner"
				);
				expect(component.length).toBe(1);
			});

			it("should match the default loading snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});

		//testing error state
		describe("when request fails", () => {
			let wrapper;
			beforeEach(() => {
				const initalValue = {
					isLoading: false,
					error: "An error has occured",
				};

				React.useState = jest.fn(() => [
					initalValue,
					jest.fn(),
				]);

				wrapper = setup();
			});
			it("should render error msg", async () => {
				const wrapper = setup();
				const component = await findByTestAttr(
					wrapper,
					"error-message"
				);
				expect(component.length).toBe(1);
			});

			it("should match the error snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});

		//testing categories state
		describe("when request is success", () => {
			let wrapper;
			beforeEach(() => {
				const initalValue = {
					...defaultState,
					orders: DummyData.singleBookOrder,
					isLoading: false,
				};

				React.useState = jest.fn(() => [
					initalValue,
					jest.fn(),
				]);

				wrapper = setup();
			});
			it("should render items", async () => {
				expect(wrapper.find(ParentOrders).length).toBe(1);
			});

			it("should match the success snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});
	});

	describe("when `childview is`", () => {
		it("false, render parent component", async () => {
			const initalValue = {
				...defaultState,
				orders: DummyData.singleBookOrder,
				isLoading: false,
			};

			React.useState = jest.fn(() => [
				initalValue,
				jest.fn(),
			]);

			const wrapper = setup();
			const component = await findByTestAttr(
				wrapper,
				"component-profile-order-parent"
			);
			expect(component.length).toBe(1);
		});
		it("true, render child component", async () => {
			const initalValue = {
				...defaultState,
				orders: DummyData.singleBookOrder,
				isLoading: false,
				childOrders: DummyData.childOrdes,
				childView: true,
			};

			React.useState = jest.fn(() => [
				initalValue,
				jest.fn(),
			]);

			const wrapper = setup();
			expect(wrapper.find(ChildOrders).length).toBe(1);
		});
	});
});
