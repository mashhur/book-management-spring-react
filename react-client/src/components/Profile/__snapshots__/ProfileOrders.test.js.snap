// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile orders component test casses when an api call is made when request fails should match the error snapshot 1`] = `ReactWrapper {}`;

exports[`Profile orders component test casses when an api call is made when request is success should match the success snapshot 1`] = `ReactWrapper {}`;

exports[`Profile orders component test casses when an api call is made while loading should match the default loading snapshot 1`] = `ReactWrapper {}`;
